MM-22044: Fix panic on web_conn send hello

agnivade opened the pull request:
<!-- Thank you for contributing a pull request! Here are a few tips to help you:

1. If this is your first contribution, make sure you've read the Contribution Checklist https://developers.mattermost.com/contribute/getting-started/contribution-checklist/
2. Read our blog post about "Submitting Great PRs" https://developers.mattermost.com/blog/2019-01-24-submitting-great-prs
3. Take a look at other repository specific documentation at https://developers.mattermost.com/contribute
-->

#### Summary
<!--
A description of what this pull request does.
-->
The core problem is here:
```go
func (h *Hub) Register(webConn *WebConn) {
	select {
	case h.register <- webConn:
	case <-h.didStop:
	}

	if webConn.IsAuthenticated() {
		webConn.SendHello()
	}
}
```

We register the connection, and then we send a hello. The registration happens in its separate goroutine. And the `SendHello` happens in the main goroutine.

So essentially what happens is, in case of extreme load, after the registration is complete, the connection gets added to the main pool. And before even the runtime has a chance to execute the next 2 lines, the buffer of the channel gets filled up from the senders that are broadcasting in the other goroutine. And once it gets filled up, we close the channel. And then when the runtime tries to run `SendHello`, boom.

#### Ticket Link
<!--
If this pull request addresses a Help Wanted ticket, please link the relevant GitHub issue, e.g.

  Fixes https://github.com/mattermost/mattermost-server/issues/XXXXX

Otherwise, link the JIRA ticket.
-->
https://mattermost.atlassian.net/browse/MM-22044

lieut-data commented on app/web_hub.go:
```diff
@@ -445,6 +457,7 @@ func (h *Hub) Start() {
 						default:
 							mlog.Error("webhub.broadcast: cannot send, closing websocket for user", mlog.String("user_id", webCon.UserId))
 							close(webCon.Send)
+							webCon.SetClosed()
```
Is there still a race condition between the `close` and setting this atomic bit?

Is it strictly necessary to close the channel at all? It appears we already support shutting down the goroutine via a `endWritePump`. If we're not using this as a signal mechanism, can we leave it to garbage collection alone?

agnivade replied:
> Is there still a race condition between the close and setting this atomic bit?

I don't believe there will be. Because this entire thing happens in a single goroutine. Even if the scheduler preempts the goroutine, before it checks for `webCon.IsClosed()`, it needs to go through `SetClosed` first.

> Is it strictly necessary to close the channel at all? 

AFAIU, yes. Because in `writePump`, we check for closing of the channel and send a close message down the websocket. And _after_ that, we close the websocket which closes the `endWritePump`. So it's not just a signal mechanism, we indeed do some cleanup work when the channel is closed.

However, this is my first time working with this part of the codebase, and you obviously have better historical context. So I am open to any improvements that you have in mind.

Pomyk replied:
If it is used in one goroutine do you need to use atomic?

agnivade replied:
Strictly speaking, we don't. But this entire thing itself is happening in a separate goroutine. So it makes me a bit more comfortable knowing that it's not going to cause any troubles later, if it gets used in some other place.

Perhaps we can unexport the method, to reduce chances of it being used from other packages. There's no need to export it.

I am not a big fan of the solution in the first place. Totally open to any other solution that you may have in mind :)

agnivade commented:
@lieut-data - Re-adding this to your queue.

agnivade commented:
Had a chat with the team about this. I will spend some additional time load testing this and writing additional tools to throughly test and have more confidence that we are not breaking anything with the change. Also added another pair of eyes (Claudio) as a reviewer.

I also plan to own the websocket part of the codebase, since there is a big ticket about re-architecting our websocket logic anyways.

agnivade commented:
Ok, so after staring at this code long enough, I figured out what was happening. My initial approach was too narrow, and I could not see the root cause at the broad level.

The core problem is here:
```go
func (h *Hub) Register(webConn *WebConn) {
	select {
	case h.register <- webConn:
	case <-h.didStop:
	}

	if webConn.IsAuthenticated() {
		webConn.SendHello()
	}
}
```

We register the connection, and then we send a hello. The registration happens in its separate goroutine. And the `SendHello` happens in the main goroutine.

So essentially what happens is, in case of extreme load, after the registration is complete, the connection gets added to the main pool. And before even the runtime has a chance to execute the next 2 lines, the buffer of the channel gets filled up from the senders that are broadcasting in the other goroutine. And once it gets filled up, we close the channel. And then when the runtime tries to run `SendHello`, boom.

So I removed all the unnecessary code and essentially brought `SendHello` to run inside the same goroutine. This guarantees that registration completes, before other senders can broadcast to it. Now it's just a simple PR small enough to review. I also removed the test because the theory on which the test was written was wrong. And it's also not possible to trigger the same code path with a unit test anymore.

But the story doesn't end here. While it fixes this panic, there is yet another send happening in websocket_router.go which is right after this register code.

```go
wr.app.HubRegister(conn)
resp := model.NewWebSocketResponse(model.STATUS_OK, r.Seq, nil)
conn.Send <- resp
```

Aaaand it panics again here. But that's a story for a different time.

Until then, we just fix one race at a time.

streamer45 commented on app/web_hub.go:
```diff
@@ -415,6 +411,16 @@ func (h *Hub) Start() {
+						model.BuildNumber,
+						webCon.App.ClientConfigHash(),
+						webCon.App.License() != nil))
+					webCon.Send <- msg
```
I think the `SendHello` function can remain, just need to unexport it and pass it the `WebConn`. 
I understand it's called only here but since we are inside a pretty nested case it would look a bit more readable.

agnivade replied:
Yeah .. my only concern is if somebody calls it from outside this goroutine. But that seems very unlikely.

lieut-data replied:
1/5, but I feel like the nesting was better than the `DO NOT call this function ... ` warning :D

agnivade replied:
Should I just remove the comment ? Or revert ? :P

lieut-data replied:
I think it's fine -- no need to churn on your part. Just expressing an opinion :)

streamer45 replied:
@lieut-data That's a a fair point. 

I thought the semantic of the function plus the fact is was unexported was enough to make it safe.

Maybe we could just move the msg generation in its own function and keep the send in there.
Something like:

```go
webCon.Send <- createHelloMsg()
```

agnivade replied:
Ooh that's a good one. :+1:

agnivade commented:
@lindy65 - This is ready to be tested. Thanks.

agnivade commented:
/update-branch

mattermod commented:
Creating a new SpinWick test server using Mattermost Cloud.

mattermod commented:
Mattermost test server created! :tada:

Access here: https://mattermost-server-pr-13799.test.mattermost.cloud

| Account Type | Username | Password |
|---|---|---|
| Admin | sysadmin | Sys@dmin123 |
| User | user-1 | User-1@123 |

mattermod commented:
Test server destroyed

agnivade commented:
/cherry-pick release-5.23
